### PR TITLE
Improve AuthorizationConfig entity structure

### LINKED ISSUE #629: Improve `AuthorizationConfig` entity structure
For now we have following structure of `AuthorizationConfig`:
```
AuthorizationConfig->Roles->Set<Role>
```
intermediate entity `Roles` is excessive here and need to be reduced.

#### How to do
1. Make `AuthorizationConfig` entity look like follows:
```
AuthorizationConfig
{
    private Set<Role> roles  = new LinkedHashSet<>();
}
```
2. Remove `org.carlspring.strongbox.security.Roles` class;
3. Resolve all compilation errors;
4. Create DB migration script to drop `AuthorizationConfig.roles` field, and replace it with `AuthorizationConfig.Set<Role>` field;
migration should be placed in `strongbox-data-service/src/main/db/changelog/v1.0.0`
5. Check that `mvn clean install` completes with `SUCESS` result.

### PR DESCRIPTION
**Pull request by ptirador**
Fixes #629.

1. `org.carlspring.strongbox.users.security.AuthorizationConfig` refactored.
2. `org.carlspring.strongbox.security.Roles` class removed.
3. Compilations errors (derived from the previous changes) resolved.
3. DB migration script created.
4. `mvn clean install` completes with SUCESS result.

### CODE REVIEW

File: strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml [hunk 1]
```diff
@@ -1,49 +1,28 @@
...
-        <!-- Will lead to ConfigurationException: Embedded roles overriding is forbidden: [ADMIN] -->
-        <!--<role>
-            <name>adMIN</name>
-            <description>Deployment role</description>
-            <privileges>
-                <privilege>Deploy</privilege>
-            </privileges>
-        </role> -->
-    </roles>
-    <privileges>
```
Comment (carlspring) [inline strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml]: So, what... We'll no longer have separate privileges...?

File: strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml [hunk 2]
```diff
@@ -1,49 +1,28 @@
...
-        <privilege>
-            <name>EXAMPLE_PRIVILEGE</name>
-            <description>This is an example of how you could add some custom privilege. Feel free to use it by name at
-                the REST API side to access restriction purposes.
-            </description>
-        </privilege>
-    </privileges>
+
+    <!-- special treatment role -->
+    <role>
```
Comment (carlspring) [inline strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml]: I believe the requirement was to drop the wrapper class, but not the actual `<roles/>` XML wrapper. Could you please change this?

File: strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/users/AuthorizationConfigProviderTest.java [hunk 3]
```diff
@@ -65,7 +65,7 @@ public void testUpdateAndAccessToEmbeddedEntities()
                                      Role testRole = new Role();
                                      testRole.setName("MY_ROLE");
                                      testRole.getPrivileges().add("MY_PRIVILEGE");
-                                     authorizationConfig.getRoles().getRoles().add(testRole);
+                                     authorizationConfig.getRoles().add(testRole);
```
Comment (carlspring) [inline strongbox-security/strongbox-user-management/src/test/java/org/carlspring/strongbox/users/AuthorizationConfigProviderTest.java]: Thanks! This looks much better! :)

File: strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml (hunk 2 above)
Comment (sbespalov) [reply strongbox-security/strongbox-user-management/src/main/resources/etc/conf/strongbox-authorization.xml]: +1

**Comment by ptirador** [comment]
The changes commented have been reverted, but in order to make work `AuthorizationConfigControllerTestIT.java` tests, I had to change the XML `strongbox-web-core/src/test/resources/etc/conf/strongbox-authorization.xml.`

File: strongbox-web-core/src/test/resources/etc/conf/strongbox-authorization.xml [hunk 4]
```diff
@@ -1,36 +1,28 @@
...
+        </privileges>
+    </role>
+    <role>
+        <name>CUSTOM_ROLE</name>
+        <description>Deployment role</description>
+        <privileges>
+            <privilege>Deploy</privilege>
+        </privileges>
+    </role>
+
```
Comment (sbespalov) [inline strongbox-web-core/src/test/resources/etc/conf/strongbox-authorization.xml]: @ptirador 
we dont want to drop wrapper `<roles>` tag here, so the xml should still looks as it was like this:
```
    <roles>
        <!-- special treatment role -->
        <role>
            <name>ANONYMOUS_ROLE</name>
            <description>Common anonymous user role</description>
            <privileges>
                <privilege>ARTIFACTS_RESOLVE</privilege>
                <privilege>SEARCH_ARTIFACTS</privilege>
            </privileges>
        </role>
        <role>
            <name>USER_ROLE</name>
            <description>Common user role</description>
            <privileges>
                <privilege>VIEW_USER</privilege>
            </privileges>
        </role>
        <role>
            <name>CUSTOM_ROLE</name>
            <description>Deployment role</description>
            <privileges>
                <privilege>Deploy</privilege>
            </privileges>
        </role>
    </roles>
    <privileges>
        <privilege>
            <name>EXAMPLE_PRIVILEGE</name>
            <description>This is an example of how you could add some custom privilege. Feel free to use it by name at
                the REST API side to access restriction purposes.
            </description>
        </privilege>
    </privileges>
```

also seems that you dropped all the `<privelegies>` here too

Comment (sbespalov) [reply strongbox-web-core/src/test/resources/etc/conf/strongbox-authorization.xml]: you can try to use ` @XmlElementWrapper(name = "roles")`  like this:
```
    @XmlElementWrapper(name = "roles")
    @XmlElement(name = "role")
    @Embedded
    private Set<Role> roles = new LinkedHashSet<>();
```

Comment (sbespalov) [reply strongbox-web-core/src/test/resources/etc/conf/strongbox-authorization.xml]: @ptirador sorry for confusion about `<privelegies>`, seems that we really should remove it as there is no such field in `AuthorizationConfig`.

**Comment by ptirador** [comment]
@sbespalov I've pushed the changes.

**Comment by steve-todorov** [comment]
@sbespalov, @carlspring ping? What's the state of this issue? :)

File: strongbox-data-service/src/main/resources/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<databaseChangeLog xmlns:o="http://www.unbroken-dome.org/schema/liquibase-orientdb"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                        http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.2.xsd">
+
+    <changeSet id="v1.0.0.8" author="pablotr87@gmail.com">
+
+        <o:dropProperty name="AuthorizationConfig.roles" className="AuthorizationConfig" propertyName="roles" />
```
Comment (sbespalov) [inline strongbox-data-service/src/main/resources/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml]: @ptirador I think sometiong wrong here :)
probably it should be:
```
<o:dropProperty name="roles" className="AuthorizationConfig"  />
```

**Comment by sbespalov** [comment]
there is one small issue with migration script, @ptirador  can you please have a look and it can be merged after this

File: strongbox-data-service/src/main/resources/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml (hunk 5 above)
Comment (ptirador) [reply strongbox-data-service/src/main/resources/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml]: What's the problem?
If I change it to that, an error occurs:
```
Caused by: liquibase.exception.ValidationFailedException: 
Validation Failed:
     1 changes have validation failures
          propertyName: must not be null, classpath:/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml::v1.0.0.8
```

Comment (sbespalov) [reply strongbox-data-service/src/main/resources/db/changelog/v1.0.0/v1.0.0.8__AuthorizationConfig.xml]: > Which one's the correct one?

CASE 4

I have also checked and it works like this:
```
        <o:dropProperty className="AuthorizationConfig" propertyName="roles" />
        <o:dropClass name="Roles" />
        <o:createProperty name="roles" type="embeddedset" linkedType="Role" className="AuthorizationConfig" />

```

**Comment by ptirador** [comment]
@sbespalov The script has been changed as requested. Could you take a look? :)

**Comment by sbespalov** [comment]
@ptirador looks good, many thanks.
@carlspring @steve-todorov I think it can be merged now

**Comment by carlspring** [comment]
Thanks for this fix and for the review! :)

**Pull request merged**